uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js

```diff
@@ -141,52 +136,138 @@ define([
         vm.loading.tab = false;
       });
     }());
 
     /**
-     * Calculate change in balance, it updates local balance variables.
+     * Calculates closing balance which is opening balance minus change amount.
+     * We use "+" operation since the change amount is negative.
+     */
+    function calculateClosingBalance () {
+      vm.balance.closing = vm.balance.opening + vm.balance.change.amount;
+    }
+
+    /**
+     * Converts given date to javascript date as expected by uib-datepicker
      *
-     * @return {Promise} empty promise if all required params are not set
-     *   otherwise promise from server
+     * @param {String} date from server
+     * @return {Date}
      */
-    function calculateBalanceChange () {
-      vm._setDateAndTypes();
-      vm._toggleBalance();
+    function convertDateFormatFromServer (date) {
+      return moment(date, sharedSettings.serverDateFormat).toDate();
+    }
 
-      if (!vm._canCalculateChange()) { return $q.resolve(); }
+    /**
+     * Converts given date to server format
+     *
+     * @param {Date} date
+     * @return {String} date converted to server format
+     */
+    function convertDateToServerFormat (date) {
+      return moment(date).format(sharedSettings.serverDateFormat);
+    }
 
-      vm.loading.showBalanceChange = true;
+    /**
+     * Sets dates (with time in case of "hours" Absence Type) from UI to request
+     *
+     * @param  {String} dateType from|to
+     * @return {Promise}
+     */
+    function dateChangeHandler (dateType) {
+      var absencePeriodChanged = isAbsencePeriodChanged(dateType);
 
-      return vm.request.calculateBalanceChange(vm.selectedAbsenceType.calculation_unit_name)
-        .then(setBalanceChange)
+      setDatepickerBoundaries();
+      (dateType === 'from') && resetToDateIfGreaterThanFromDate();
+      resetUIDayTypesTimeAndDeductions(dateType);
+
+      vm.uiOptions.times[dateType].loading = true;
+      vm.loading[dateType + 'DayTypes'] = true;
+      vm.uiOptions.times[dateType].disabled = true;
+
+      return $q.resolve()
+        .then(vm.onDateChangeExtended)
+        .then(function () {
+          if (absencePeriodChanged) {
+            vm.period = getAbsencePeriod(dateType);
+          }
+
+          if (!vm.period.id) {
+            return $q.reject('Please change date as it is not in any absence period');
+          }
+        })
+        .then(loadCalendar)
+        .then(function () {
+          return $q.all([
+            isCalculationUnit('hours') && loadTimeRangesFromWorkPattern(dateType),
+            loadDayTypesForDate(vm.uiOptions[dateType + 'Date'], dateType)
+          ]);
+        })
+        .then(function () {
+          setRequestDateTimesAndDateTypes();
+
+          if (absencePeriodChanged) {
+            setDatepickerBoundaries();
+            $rootScope.$broadcast('LeaveRequestPopup::absencePeriodChanged');
+          } else {
+            return performBalanceChangeCalculation();
```

Comment:
I refactored the function, not 100% sure it's correct, but it makes the function easier to follow and addresses the points a raised above (unless i misunderstood the flow you are trying to have)

```js
function dateChangeHandler (dateType) {
  var absencePeriodChanged;

  return $q.resolve()
    .then(function () {
      setDatepickerBoundaries();
      resetUIDayTypesTimeAndDeductions(dateType);

      if (dateType === 'from') {
        resetToDateIfGreaterThanFromDate();
      }

      vm.uiOptions.times[dateType].loading = true;
      vm.loading[dateType + 'DayTypes'] = true;
      vm.uiOptions.times[dateType].disabled = true;
    })
    .then(vm.onDateChangeExtended)
    .then(function () {
      return getAbsencePeriod(dateType)
    })
    .then(function (datePeriod) {
      absencePeriodChanged = datePeriod.id !== vm.period.id;

      if (!absencePeriodChanged) {
        return;
      }

      vm.period = datePeriod;

      return loadCalendar();
        .then(function () {
          return $q.all([
            isCalculationUnit('hours') && loadTimeRangesFromWorkPattern(dateType),
            loadDayTypesForDate(vm.uiOptions[dateType + 'Date'], dateType)
          ]);
        });
    })
    .then(setRequestDateTimesAndDateTypes)
    .then(function () {
      if (absencePeriodChanged) {
        $rootScope.$broadcast('LeaveRequestPopup::absencePeriodChanged');
      } else {
        return performBalanceChangeCalculation();
      }
    })
    .catch(handleError);
  }
}
```

Reply:
@AkA84 this flow is not right:

```js
return loadCalendar();
        .then(function () {
          return $q.all([ // this should not be dependant on the absence period. it is dependant on the date change.
            isCalculationUnit('hours') && loadTimeRangesFromWorkPattern(dateType),
            loadDayTypesForDate(vm.uiOptions[dateType + 'Date'], dateType)
          ]);
        });
```

Reply:
However, I removed the `isAbsencePeriodChanged` function and applied a part of your suggestion.